cmd/prune-junit-xml/logparse/logparse.go

```diff
@@ -0,0 +1,221 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package logparse provides a parser for the klog text format:
+//
+//	 I1007 13:16:55.727802 1146763 example.go:57] "Key/value encoding" logger="example" foo="bar" duration="1s" int=42 float=3.14 string="hello world" quotedString="hello \"world\"" multiLinString=<
+//		hello
+//		world
+//	  >
+//	  E1007 15:20:04.343375 1157755 example.go:41] Log using Errorf, err: fail
+//
+// It also supports the klog/ktesting unit test log output:
+//
+//	  === RUN   TestKlogr
+//		    example_test.go:45: I1007 13:28:21.908998] hello world
+//		    example_test.go:46: E1007 13:28:21.909034] failed err="failed: some error"
+//		    example_test.go:47: I1007 13:28:21.909042] verbosity 1
+//		    example_test.go:48: I1007 13:28:21.909047] main/helper: with prefix
+//		    example_test.go:50: I1007 13:28:21.909076] key/value pairs int=1 float=2 pair="(1, 2)" raw={"Name":"joe","NS":"kube-system"} slice=[1,2,3,"str"] map={"a":1,"b":2} kobj="kube-system/joe" string="hello world" quotedString="hello \"world\"" multiLineString=<
+//		        	hello
+//		        	world
+//		         >
+//		    example_test.go:63: I1007 13:28:21.909085] info message level 4
+//		    example_test.go:64: I1007 13:28:21.909089] info message level 5
+//	  --- PASS: TestKlogr (0.00s)
+//	  PASS
+//	  ok  	k8s.io/klog/v2/ktesting/example	(cached)
+//
+// Arbitrary indention with space or tab is supported. All lines of
+// a klog log entry must be indented the same way.
+package logparse
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"iter"
+	"regexp"
+	"strings"
+)
+
+// Parse splits log output provided by the reader into individual log entries.
+// The original log can be reconstructed verbatim by concatenating these
+// entries. If the reader fails with an error, the last log entry will
+// capture that error.
+func Parse(in io.Reader) []Entry {
+	var log []Entry
+	for entry := range All(in) {
+		log = append(log, entry)
+	}
+	return log
+}
+
+// All is like Parse except that it can be used in a for/range loop:
+//
+//	for entry := range logparse.All(reader) {
+//	    // entry is the next log entry.
+//	}
+func All(in io.Reader) iter.Seq[Entry] {
+	return func(yield func(Entry) bool) {
+		parse(in, yield)
+	}
+}
+
+// Entry is the base type for a log entry. It's String implementation
+// returns the original log text, including the line break.
+//
+// Use type assertions to check for additional information. All
+// of the instances behind this interface are pointers.
+type Entry interface {
+	String() string
+}
+
+// ErrorEntry captures the error encountered while reading the log input.
+type ErrorEntry struct {
+	Error error
+}
+
+func (e *ErrorEntry) String() string { return "" }
```

Comment:
didn't look yet to the whole logic, but why return an empty string instead of the error message?

Reply:
Because the API says that concatenating the string representation of all entries recreates the input text.

The error wasn't part of that input, so it shouldn't be turned into one.

This `ErrorEntry` was necessary to support the new `for/range` pattern in Go 1.23. The code producing new items to iterate over has no way to return errors. I picked that pattern because I wanted to experiment a bit with it, not because it is really needed.

Reply:
oh, I see, https://go.dev/blog/range-functions you always find some use for the cool things 😄 

I need to educate myself then

Reply:
I probably should rename `Entry.String` to `Entry.Log`, then this `String` can return the error.

Reply:
Or `LogData`.

Reply:
I don't have enough time to review the parser, it looks that the complexity it adds vs a normal parsing is significant, but  I fully trust you and dims ... I left some comments on the other logic though

Reply:
thanks @aojea ... to be fair, We'll invariably (hopefully not!) find edge cases, so we have to keep vigilant when we are looking at the pruned logs. Since this is not going into a release itself per-se, we have a bit more lee way.

Reply:
yeah, I just wanted to mean that I'm not worried at all, because of what you say 😄